model/wire.cc

```diff
@@ -137,12 +137,125 @@ namespace minsky
       dest.addWire(wp);
   }
 
+// For ticket 991. Adapted from https://www.stkent.com/2015/07/03/building-smooth-paths-using-bezier-curves.html
+namespace
+{
+	
+  vector<float> constructLowerDiagonalVector(int length) {
+    vector<float> result(length);
+
+    for (int i = 0; i < result.size() - 1; i++) {
+      result[i] = 1.0;
+    }
+    
+    result[result.size() - 1] = 2.0; 
+
+    return result;
+  }
+
+  vector<float> constructMainDiagonalVector(int n) {
+    vector<float> result(n);
+
+    result[0] = 2.0;
+
+    for (int i = 1; i < result.size() - 1; i++) {
+      result[i] = 4.0;
+    }
+    
+    result[result.size() - 1] = 7.0; 
+
+    return result;
+  }
+
+  vector<float> constructUpperDiagonalVector(int length) {
+    vector<float> result(length);
+
+    for (int i = 0; i < result.size(); i++) {
+      result[i] = 1.0;
+    }
+
+    return result;
+  }	
+  
+  vector<pair<float,float>> constructTargetVector(int n, const vector<pair<float,float>> knots) {
+	
+	assert(knots.size() > 2);
+	  
+    vector<pair<float,float>> result(n); 
+
+    result[0].first = knots[0].first+2.0*knots[1].first;
+    result[0].second = knots[0].second+2.0*knots[1].second; 
+    
+    for (int i = 1; i < n - 1; i++) {
+      result[i].first = 2.0*(2.0*knots[i].first+(knots[i+1].first)); 
+      result[i].second = 2.0*(2.0*knots[i].second+(knots[i+1].second));  
+    }
+    
+    result[result.size() - 1].first = 8.0*knots[n-1].first+knots[n].first; 
+    result[result.size() - 1].second = 8.0*knots[n-1].second+knots[n].second;
+
+    return result;
+  }  
+
+ vector<pair<float,float>> computeControlPoints(int n, const vector<pair<float,float>> knots) {
```

Comment:
I feel this method needs more documentation than just a link to a web page, that may not exist at some point in the future. It is possible to embed LaTeX into Doxygen comments, so don't feel restricted to plain ASCII.

The code feels like it can be expressed via some matrix equation, where the matrix is tridiagonal.

Reply:
Thank you, that is a very helpful tip. Based on it, I have found this very well-documented implementation of Thomas's algorithm for solving the matrix equation Ac = r where A is nearly tridiagonal: https://github.com/CodyJae/Thomas-Algorithm/blob/master/Thomas-Algorithm.cpp. In our case, the matrix is fully tridiagonal, so the implementation will be simpler than at the aforementioned Github page.

Am I allowed to use Doxygen in .cc files? I've only seen it in the header files. For the documentation using LaTeX in Doxygen, https://en.wikibooks.org/wiki/Algorithm_Implementation/Linear_Algebra/Tridiagonal_matrix_algorithm#C++ looks like a good template.

I will implement the above and remove all the dead code.

Reply:
Yes - you can use Doxygen in .cc files. Whether doxygen processes .cc files or not is a configuration option. Embedding LaTeX into comments in a .cc file is a reasonable use case. Cross referencing to other parts of the code base is another reason.

But for comments about interface related matters, it is better to please these in the header files - curiously in a competing documentation processor called gtk, the tradition is to add the documentation to the .c files (its mostly used for C-based code, not C++), rather than .h. They have their reasons, but I find it annoying coming across a header file in such a project, to find no useful comments whatsoever, and of course the C files have all disappeared into an object code library.

Reply:
Re what needs to go here, - we don't need a treatise. I would basically mention that two control points are added in between each handle, and the curves and first and second derivatives matched, second derivatives set to zero at the ends to absorb the extra degrees of freedom, giving rise to to that matrix equation (can copy the one from the blog you refer to). Then Thomas algorithm used to solve the matrix equation. Probably only about 2 paragraphs max, but enough bread crumbs that that algorithm can be reconstructed or analysed without access to the blog article.

Reply:
Okay, I have replaced the three vector<pair<float>> functions to construct the upper diagonal, diagonal, and lower diagonal vectors with a single vector<vector<float>> matrix function. I tried to implement the single function at the Github page I linked to in my last reply, but could not manage to get it to work. 

The current implementation now calls two anonymous namespace functions in Wire::draw(), which respectively construct the matrix A and target knots k (in Ac = k), and sends c (the coordinates of all the handles), A and k to an anonymous namespace function computeControlPoints(A,c,k). The latter, in turn, applies Thomas' algorithm to Ac =k and returns the new control points. 

All that remains to be done is to add the Doxygen documentation, which I will do today. Thanks for the outline of what it should look like. :)

Reply:
I've added the Doxygen documentation in wire.cc and fixed all LaTeX related errors that came up when running doxygen from the source directory. The pdf file generated under the latex directory of doxydoc is mostly empty though...